File: src/main/python/com/example/tutorial/dsa/medium/concurrency/counter/counter.py
"""
Counter Protocol - Interface for thread-safe counters.

This protocol defines the contract that all counter implementations must follow.
Python's Protocol is similar to Java's interface.
"""

from typing import Protocol


class Counter(Protocol):
    """Protocol for counter implementations."""

    def increment(self) -> None:
        """Increments the counter by 1."""
        ...

    def get_count(self) -> int:
        """Returns the current count."""
        ...

File: src/main/java/com/example/tutorial/dsa/medium/concurrency/counter/Counter.java
package com.example.tutorial.dsa.medium.concurrency.counter;

/**
 * Counter interface for thread-safe counter implementations.
 *
 * <p>Defines the contract for counter operations used to demonstrate
 * different thread-safety approaches in concurrent programming.
 */
public interface Counter {
  /**
   * Increments the counter by 1.
   */
  void increment();

  /**
   * Gets the current counter value.
   *
   * @return current count
   */
  int get();
}
